indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java

```diff
@@ -184,275 +136,388 @@ public String getType()
+                                                delegateFirehoseFactory,
+                                                ingestionSchema.getTuningConfig().isReportParseExceptions(),
+                                                null, null
+                                            );
+    try {
+      final Map<Interval, List<ShardSpec>> shardSpecs = determineShardSpecs(toolbox, firehoseFactory);
 
-    final Set<Interval> validIntervals = Sets.intersection(granularitySpec.bucketIntervals().get(), getDataIntervals());
-    if (validIntervals.isEmpty()) {
-      throw new ISE("No valid data intervals found. Check your configs!");
-    }
+      final String version;
+      final DataSchema dataSchema;
+      if (!determineIntervals) {
+        version = Iterables.getOnlyElement(getTaskLocks(toolbox)).getVersion();
+        dataSchema = ingestionSchema.getDataSchema();
+      } else {
+        Interval interval = JodaUtils.umbrellaInterval(shardSpecs.keySet());
+        TaskLock lock = toolbox.getTaskActionClient().submit(new LockAcquireAction(interval));
+        version = lock.getVersion();
+        dataSchema = ingestionSchema.getDataSchema().withGranularitySpec(
+            ingestionSchema.getDataSchema()
+                           .getGranularitySpec()
+                           .withIntervals(
+                               JodaUtils.condenseIntervals(
+                                   shardSpecs.keySet()
+                               )
+                           )
+        );
+      }
 
-    for (final Interval bucket : validIntervals) {
-      final List<ShardSpec> shardSpecs;
-      if (targetPartitionSize > 0) {
-        shardSpecs = determinePartitions(bucket, targetPartitionSize, granularitySpec.getQueryGranularity());
+      if (generateAndPublishSegments(toolbox, dataSchema, shardSpecs, version, firehoseFactory)) {
+        return TaskStatus.success(getId());
       } else {
-        int numShards = ingestionSchema.getTuningConfig().getNumShards();
-        if (numShards > 0) {
-          shardSpecs = Lists.newArrayList();
-          for (int i = 0; i < numShards; i++) {
-            shardSpecs.add(new HashBasedNumberedShardSpec(i, numShards, null, jsonMapper));
-          }
-        } else {
-          shardSpecs = ImmutableList.<ShardSpec>of(NoneShardSpec.instance());
-        }
+        return TaskStatus.failure(getId());
       }
-      for (final ShardSpec shardSpec : shardSpecs) {
-        // ShardSpec to be published.
-        final ShardSpec shardSpecForPublishing;
-        if (ingestionSchema.getTuningConfig().isForceExtendableShardSpecs()) {
-          shardSpecForPublishing = new NumberedShardSpec(
-              shardSpec.getPartitionNum(),
-              shardSpecs.size()
-          );
-        } else {
-          shardSpecForPublishing = shardSpec;
+    }
+    finally {
+      if (Closeable.class.isAssignableFrom(firehoseFactory.getClass())) {
```

Comment:
A design goal here was to keep the existing Firehose and FirehoseFactory interfaces without modification. So the first time a user calls `ReplayableFirehoseFactory.connect()` it creates a new firehose which reads the delegate firehose into cached files. Once the ReplayableFirehose is read through, it's expected the user will close the firehose which will close the filestream *but* leaves the cached files on disk (obviously, that's the whole point). 

The ReplayableFirehoseFactory maintains a reference to the firehose it created so when `connect()` is called again, it returns the original firehose *"rewound"*, and doesn't read from the delegate again.

So the purpose of closing the factory is really just to delete those temporary files to free up disk space early. The files are all marked with `deleteOnExit()` so forgetting to close the factory isn't fatal, but closing it is nicer.